### DISCUSSION TITLE
Perfera heating logic

### DISCUSSION THREAD

**Discussion by miniluma** [revk/xESP32-Faikout#226] [General]
Hi, first of all, I really love this project! Great work.

Regarding Perfera heating logic, I have a problem with heatover. I noticed an error when setting negative values for heatover. ("must be positive").

Here is how Perfera wall mounted units work in heat mode (This is proven. I can provide more details an links, if you want):
- There is an offset of the measured room temp / inlet teamp to the real room temp. (Probably not an offset, but caused by to the construction of the Perfera units)
- The Daikin logic stops the heating process, if the delta temp between target temp and inlet/room temp >= 4K
- E.g.: Target temp of the unit is set to 20.0°C. The unit started heating. If the measured temp is 23.5 everything is fine and heating goes on. As soon as the measured temp equals 24, the unit goes off.
- Depending on the deltaT, the unit tries to provide more heat to the room. DeltaT = 2 or 2.5 is good. If it comes closer to 4, it reduces the power of the outdoor unit and after a while also the fan speed. DeltaT < 2 leads to higher liquid temperatures. This helps to heat up cold rooms, but is inefficient.

So the optimal case for a long-running system with very low power consumption, a good SCOP (aka. low liquid temps) and low or no on/off cycles is to manage the deltaT.

Your current heating logic (with _heatover=switchtemp=pushtemp=0_) makes _set = max + reference - current_ which is like _inlet+1_. (heatover and swichtemp increase is, right?)
But heating will start already at inlet-3.5. So you are requesting a lot of heat from the system

My first proposal: Please allow negative values for heatover 
My second proposal:
Adapt the logic to Perfera behavior (at least as an option)
E.g. at current < min, set the unit target temp to inlet-2.
Let the algorithm readjust this as soon as inlet changes (always _set=inlet-2_)
If the room temp comes close to the desired temp (e.g. desired-0.1) reduce to inlet-3 (and follow any inlet changes)

As an experiment, you could set the target temp to inlet-3.5 if the desired temp is reached and keep it until _current>max_.
(It depends how quick you get the inlet in S21 to avoid the deltaT=4K and how fast the indoor unit deactivates if deltaT=4K. That needs to be tested)

**Comment by revk** [comment]
Do we think there is a bug, or do I just need to make these signed variables?

**Reply by turbolooser** [reply]
In my case I think its a bug so I have raised https://github.com/revk/ESP32-Faikin/issues/227
and removed it from here

**Reply by miniluma** [reply]
Let me think about it. Maybe it needs more logic...

**Comment by miniluma** [comment]
@revk Let me show you what I mean by DeltaT Management:
This the what happens, when I make _setTemp=inlet-3.5_
![image](https://github.com/revk/ESP32-Faikin/assets/77453740/95402e08-5c46-47f9-bf05-edb7f12ef02d)
Red is the settemp, black is the inlet, and green is a homematic room thermostat.
This is a 2MXM40A. Both rooms on quite/night mode. Demand 40%.
The lower room takes the live status of home and sends temp command accordingly. (consequently home-3.5).
The upper room has no faikin yet, so it does the same on the values from the Daikin Cloud. (Works similar, but with delayed values of ~1Min).

You see pretty steady inlet temps and the room is keeping it's temperature and I also can cool it down a bit (this only works for cooler outdoor temperatures. it's 5°C this morning)
liquidtemp is at 25-26°C (which is 10K below running it with the current faikin logic and much more energy efficient).
Also, the outdoor unit has 120-180W consumption while heating or keeping 2 rooms (blue line).
![image](https://github.com/revk/ESP32-Faikin/assets/77453740/4b9965f7-4db1-47c0-a868-958a69cddce8)

So, to conclude: Your faikin logic isn't wrong, let me say that at first.
But it's also not using the full potential to control and keep the room temperature and avoid heatback mode. So overshooting to quickly heat cold rooms is good. When you nearly hit the target room temperature, you could reduce the daikin set temperature much lower (up to -3.5, the lowest I've seen with heatover=0 and switchtemp=0 was settemp==inlet)

**Reply by revk** [reply]
Thanks, I originally did try and make the target temp I send more dynamic, closer when closer to the desired temperature, but i found I was fighting the logic in the Daikin. I found asking it to slightly increase temperature it would decide it is close enough and do nothing, etc. It may be that other models are better at that and a different logic may be needed. I hoped that I have enough settings to allow for various ways of working.

I am not against making some alternative algorithms and settings, but if my unit won't play it will make testing very difficult here.

If you look at the code, you may be able to make some specific suggestions for changes.

**Reply by miniluma** [reply]
I agree. It's always fighting the Daikin Algorithm.
I already cloned the repo but I'm not able to build yet.
I'm sure I will come back with some questions as soon as it works.

Btw: Are you aware of the Daikin brc073? I'm wondering if it sends a room temperature value to the indoor unit.

**Reply by revk** [reply]
If you come up with ideas, let me know, if you want to do a pull request, please make different algorithm based on a setting!

**Reply by turbolooser** [reply]
@miniluma sounds great.... looking forward

The brc073 uses the S21 port also. So there needs to be a way to get the external temp used as thermal ref to the daikin

**Reply by revk** [reply]
The faikin can be set up to snoop on messages sent (tap the rx line on to what you want to monitor and fix the protocol), if you are able to snoop the traffic and find how it does it, that could be useful.

**Reply by njh** [reply]
> Btw: Are you aware of the Daikin brc073? I'm wondering if it sends a room temperature value to the indoor unit.

I bought a [BRC073A1](https://www.daikin.co.uk/en_gb/products/product.html/BRC073.html) on eBay to see if this was possible.
There is a component on it that looks a lot like some kind of temperature sensor bead, and a vent in the case, so that air can get to it:
<img src="https://github.com/revk/ESP32-Faikin/assets/38734/038d5cc8-b4d0-442e-a857-03358cc393a3" width="400" />

My progress so far is to connect the Wall Controller up to the Faikin S21 simulator. And I can see it sending F2 and F8 commands. However I have not been able to get it into full operation, it is currently stuck on this screen:

<img src="https://github.com/revk/ESP32-Faikin/assets/38734/46a1920e-a3a6-4db0-9d9f-e71cb9c1f76d" width="400" alt="Checking the connection. Please standby." />

So there must be something that the simulator needs to send to it.


Slightly disappointingly, there is a statement in the [installation manual](https://www.daikin.co.uk/content/dam/document-library/installation-manuals/ctrl/BRC073_IOM_4PEN392190-1_Installation%20manuals_English.pdf) that says:

> **NOTICE**: The room temperature is measured inside the indoor unit.

So it is possible that this isn't going to work with the BRC073.

**Reply by revk** [reply]
Ah, I have one of those type of controllers, did not think they did S21, or is that what the other box does.

However, I never got it to actually *use* that temp sensor. Try freezer spray on it and see.

**Reply by njh** [reply]
There is a YouTube video here that explains how to reconfigure it to activate the temperature sensor in the Wall Controller:

https://www.youtube.com/watch?v=b34upd8EpK8&t=23s

But it is for an older model (BRC1E61 / BRC1E62).

I guess that would be an obvious next step - to reconfigure the wall controller and see if the OneCTA app can see the temperature changing on the wall controller bead.

**Reply by revk** [reply]
If it does, then we definitely want to try and monitor that S21 link!

**Reply by turbolooser** [reply]
If I get one BRC073 could I connect it somehow to the Daikin and in parallel to the Faikin to get the required messages dumped/collected etc.?
If that somehow sounds possible I am eager to get one... 

@njh Do you have a Daikin indoor unit also? Or you don't and thats why you run the S21 simulator?

**Reply by revk** [reply]
I have 3 ducted (X50A) and one call mount (S21).

The Faikin can be set to monitor one way, its `rx` connected to the line you want to monitor, `snoop` set. It then logs on MQTT what it sees. Which two Faikin you can monitor both ways.

**Reply by miniluma** [reply]
@njh Wow! That's cool! What was the price for the BRC073?

**Reply by turbolooser** [reply]
On ebay germany you can fid them for 60 bucks :-) I will get one for sure now.

**Reply by turbolooser** [reply]
Once I have the BRC073 I will install it to S21 and need your help then on the faikin connection to S21 to monitor for any messages. Will let you know

**Reply by turbolooser** [reply]
And to me this clearly says its having a temp sensor in the wall unit 
<img width="337" alt="image" src="https://github.com/revk/ESP32-Faikin/assets/115738363/460282b1-a659-4961-afee-808d5839689c">

**Reply by miniluma** [reply]
Nice!
Btw, we could think about using a second faikin right behind the BRC073 and create a S21 Wifi-bridge, so that you don't have to install the cable between the BRC073 and the indoor unit. Should not be too complicated to implement from a software perspective... Just like a S21-forward-mode ;-)

**Reply by turbolooser** [reply]
That sounds even greater. So by this we would create a wireless wall unit ? whau...
So lets see how this goes by cable as step 1

**Reply by njh** [reply]
I have had a quick look at the BRC073 in my lunch break.

Unfortunately it doesn't seem to use the thermistor in the BRC073 by default.
I have tried warming up the thermistor bead but the indoor temperature in the OneCTA app does not change.
(I can't work out how to get the BRC073 to display the indoor temperature)


I have taken a look at the Field Setting menu and Mode `20` (as shown in the YouTube video) isn't available. The only Modes that can be selected are: `1b`, `1c` and `1e`

These are the settings that are available:

| Mode | Settings                                  |
| ---- | ----------------------------------------- |
| 1b   | 8=02, 11=01, 13=02, 15=01                 | 
| 1c   | 0=01, 2=02, 4=03, 5=01, 6=01              | 
| 1e   | 1=01, 2=04, 5=02, 6=02, 8=01, 9=02, 10=01 | 

I found a service manual, for the VRV4 indoor units that mentions modes `1b` / `1c` / `1e`:

http://www.daikintech.co.uk/Data/VRV-Indoor/FXFQ/2018/FXFQ-AVEB/SM_ESIE15-11B.pdf


It says that in mode `1c`, setting `1` controls which thermistor is used:

```
Selection air thermistor shown in room temp display:
01: air return (R1T)
02: thermistor of BRC1E
```

Sadly setting `1` is not selectable on my BRC073.

![IMG_9429](https://github.com/revk/ESP32-Faikin/assets/38734/1a338e03-8137-443b-8598-253f4b9ad821)

**Reply by miniluma** [reply]
Are you able to sniff the S21 messages? If yes, can you try to set the fan to a fixed level and then set it to night mode in the app and see what messages the A/C sends to the remote?

**Reply by miniluma** [reply]
@njh What kind of cable do you have between the BRC073 and the little box?

**Reply by njh** [reply]
Sorry I missed you message

> Are you able to sniff the S21 messages?

I have successfully connected the BRC073 to my computer and I am able to send messages to and from it, using the [Faikin S21 simulator](https://github.com/revk/ESP32-Faikin/blob/main/Tools/faikin-s21.c).

> If yes, can you try to set the fan to a fixed level and then set it to night mode in the app and see what messages the A/C sends to the remote?

I am not sure a wall controller is needed for this?
You could just do this using the Onecta App or the IR Remote control and then dump messages using Faikin over MQTT?


> What kind of cable do you have between the BRC073 and the little box?

I decided not to buy an official cable. There is a JST-ZH-4 connector on both ends, so I bought a short assembled cable on eBay:

https://www.ebay.co.uk/itm/331897429718

**Reply by njh** [reply]
Just tried playing with the wall controller and I don't think it has 'Night Mode'.

Pressing the Fan button and then changing the levels (Up/Down) jumps between:
* Auto
* Lv. 1 (L)
* Lv. 2
* Lv. 3 (M)
* Lv. 4
* Lv. 5 (H)

**Reply by njh** [reply]
Just sniffed what the Wall Controller is sending out for each fan setting.
(this is byte 4 of the payload for the D1 command)

| Option    | Ascii | Hex  |
| --------- | ----- | ---- |
| Auto      | 'A'   | `41` |
| Lv. 1 (L) | '3'   | `33` |
| Lv. 2     | '4'   | `34` |
| Lv. 3 (M) | '5'   | `35` |
| Lv. 4     | '6'   | `36` |
| Lv. 5 (H) | '7'   | `37` |

**Reply by revk** [reply]
That's what we send from Faikin...

**Reply by njh** [reply]
Yeah, I should have said that - no new discoveries.

**Reply by serimi89** [reply]
Thanks for all this testing @njh - just catching up on the thread.
Have you tried changing the available field settings while observing the data comms on the S21 port? Maybe other field settings are writeable, but not displayed (wild guess, I know 😄)

**Reply by njh** [reply]
Yes, good idea! I will try and spend some time on this at the weekend.
I wonder if all Field Settings result in a S21 command being sent, or if some are local to the BRC073

**Reply by davea** [reply]
Have you tried sniffing S21 data while the remote is plugged into the wall mount unit? I wonder if the unit can issue supported features to the remote to control its behaviour?

**Reply by njh** [reply]
I don't currently have an easy way of doing that. I could probably connect up a Faikin RX line and the BRC073 at the same time - but then I will only be able to monitor either what the BRC073 is transmitting or receiving - not both at the same time.

However, I have been monitoring what my indoor unit has been transmitting and receiving using Faikin and then putting those values into the Simulator. And then sending the messages that the BRC073 is sending to the simulator and sending those using Faikin. So I should be pretty close to monitoring what the BRC073 would be doing if it was connected directly.

I have been seeing a lot of `F2`, `D2` and `F8` commands, being sent by the BRC073. I think it is currently unknown what these do.

**Reply by revk** [reply]
Are you in the UK? I have some special double sided test Faikin boards for this, and could post one.

![322559](https://github.com/revk/ESP32-Faikin/assets/996983/1be00473-4151-4503-b4c1-425b989dfdc6)
![322560](https://github.com/revk/ESP32-Faikin/assets/996983/e84e1ec6-4fa0-4cb1-8710-a1b71afef047)

**Reply by davea** [reply]
I've ordered a BRC073 and cables from eBay so would also be interested in a two-way Faikin if you've got one to spare!

**Reply by revk** [reply]
If in UK I can easily post, they are made for this very case, I have not done the monitoring s/w yet, but that is pretty simple to write. Just need an address.

**Reply by davea** [reply]
In the UK yep, not far up the road from you in fact. What's the best way to get my address to you?

**Reply by revk** [reply]
I knew you would ask that! GitHub does not do DMs does it and I try not to post my email? I'm on mastodon/fesiverse @revk@toot.me.uk

**Comment by miniluma** [comment]
@revk When I try to build the software, I end with "_ninja: error: '/home/user/ESP32-Faikin/ESP32/secure_boot_signing_key_ecdsa.pem', needed by 'signature_verification_key.bin', missing and no known rule to make it_"
Do you have a proposal how I could build a test software for my changes without much effort?

**Reply by revk** [reply]
You need to make your signing key... ESP IDF has instructions for that. Or turn off code signing.

**Reply by miniluma** [reply]
When I turn it off, can I make OTA updates of my builds and is it possible to come back to your releases later on by just changing the OTA host?

**Reply by revk** [reply]
If you have code signing off your boot loader should be able to load my code via ota. Going back may need you to reflash as my code won't ota your code unsigned.

**Comment by turbolooser** [comment]
Any Insights on this new setting? :-D 
What will it do?

<img width="594" alt="image" src="https://github.com/revk/ESP32-Faikin/assets/115738363/252bfbdf-62ce-4557-95b6-ded4ecd5b28c">

**Reply by revk** [reply]
FFS, this is what I get for allowing a pull request, no clue. I'll check.

**Reply by miniluma** [reply]
@revk Would it be better to move to a FTXMTesting branch to avoid unwanted changes in your main? Probably, you could provide some testing builds from time to time?

**Comment by miniluma** [comment]
> Any Insights on this new setting? :-D What will it do?
> 

Nothing yet. Going to try to implement a kind of special logic for the Perfera.

**Comment by revk** [comment]
He will not be doing anything on this branch, it will be a merge later, if it makes sense. And we have a beta test process as well.

**Reply by MUN0X** [reply]
I have a FTXM as well (that will get a faiking installed soon) where I was struggling with the -4°C in the logic of my existing automation as well.
I'm willing to beta test if needed

But honestly do we need a device specific setting in the faikin main branch?
Is it not an option to make the existing logic more granular?
- allow negative offsets for the existing setpoint calculation (eg. heatover -3 in case as described by TS)
- Introduce a second setpoint calculation that looks solely at the current inlet temperature + a new offset (new offset -2 for TS)
 ==> take lowest value of both both setpoint calculations (in case of heating) 

This should result in the desired behavior without adding a lot of device specific changes to the code. 
Thus allowing anyone to play around with it.

**Reply by revk** [reply]
I would rather not have model specific flag (and have removed).

Yes, changing logic to allow more options, and feature specific flags, I am happy with.

**Reply by MUN0X** [reply]
Would you be willing the implement the suggestion i made?

I have a few other ideas that might be worth considering. 
How would you like me tho raise them for you? start a new topic or are feature requests via the issues tab more convenient?

**Reply by revk** [reply]
Sorry, whichever, just been a bit busy. Changing to signed is easy!

**Reply by MUN0X** [reply]
Ok, I'll issue multiple suggestions in the coming days via multiple feature requests. 
I hope that helps to keep them separated for discussion and follow up. (or for you to dismiss them :-) )

**Comment by miniluma** [comment]
@revk I made a first shot of FTXM Control logic. 
All depending on the isFTXM setting. 
If it's set to 0, this should not affect your code.
If it's set to 1, this disables a lot of logic and applies a simple heating logic.
I also added some test to Advanced.md including the limitations.

See https://github.com/revk/ESP32-Faikin/compare/main...miniluma:ESP32-Faikin:ftxm_testing

Do you want me to make a PR or would you build a testing version so that I can test it first?

**Reply by turbolooser** [reply]
If you are looking for testers ... 👍

**Reply by miniluma** [reply]
If you want to test it, you can clone from [here](https://github.com/miniluma/ESP32-Faikin/tree/ftxm_testing) and build it.
My build fails due to some include stuff :-/

**Comment by sam-m7** [comment]
I've just discovered this thread and wanted to add some information. It's possible to change the 4 degrees offset of the Perfera units using the remote. https://www.manualslib.com/manual/2160120/Daikin-Perfera-Rxm-R-Series.html?page=251#manual

**Reply by Raspdomo** [reply]
You can indeed change it on the indoor units with the remote, however I prefer to do it in Domoticz.
